gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java

```diff
@@ -565,6 +570,64 @@ public FSDataInputStream open(Path hadoopPath, int bufferSize) throws IOExceptio
 
     return new FSDataInputStream(in);
   }
+  // Conf not used in the method for sometime, as some changes need to be made for laters
+  private void initThreadPools() {
+    int maxThreads = DEFAULT_MAX_THREADS.getDefault();
+    if (maxThreads < 2) {
+      logger.atWarning().log("Maximum number of threads should at least be 2.");
+      maxThreads = 2;
+    }
+    int keepAliveTime = KEEP_ALIVE_TIME.getDefault();
+    unboundedThreadPool =
+        new ThreadPoolExecutor(
+            maxThreads,
+            Integer.MAX_VALUE,
+            keepAliveTime,
+            TimeUnit.SECONDS,
+            new LinkedBlockingQueue<>(),
+            BlockingThreadPoolExecutorService.newDaemonThreadFactory("gcs-transfer-unbounded"));
+    unboundedThreadPool.allowCoreThreadTimeOut(true);
+  }
+
+  @Override
+  public CompletableFuture<FSDataInputStream> openFileWithOptions(
+      final Path rawPath, final OpenFileParameters parameters) throws IOException {
+    final Path path = makeQualified(rawPath);
+    Set<String> mandatoryKeys = parameters.getMandatoryKeys();
+    AbstractFSBuilderImpl.rejectUnknownMandatoryKeys(
+        mandatoryKeys, Collections.emptySet(), "for " + path);
+    StorageResourceId storageResourceId = StorageResourceId.fromUriPath(rawPath.toUri(), false);
+    GoogleCloudStorageItemInfo itemInfo = this.getGcsFs().getGcs().getItemInfo(storageResourceId);
+    checkNotNull(itemInfo, "Item info cannot be null");
+    logger.atFine().log("ItemInfo :%s", itemInfo);
+    logger.atFine().log("File exists: %s", itemInfo.getResourceId());
+    CompletableFuture<FSDataInputStream> result = new CompletableFuture<>();
+    try {
+      return unboundedThreadPool
+          .submit(() -> LambdaUtils.eval(result, () -> open(itemInfo, parameters.getBufferSize())))
+          .get();
+    } catch (InterruptedException e) {
+      e.printStackTrace();
+    } catch (ExecutionException e) {
+      e.printStackTrace();
```

Comment:
We should re-throw exception instead of ignoring it.

Reply:
Ok, will do that